Fix skipped mode-bar links & navigation order with the tab key #4095

Ifycode opened the pull request:
**This pull request makes the following changes:**
- Fixes ushahidi/platform#4095

**Issues detected:**
- Some links are skipped and when navigating the links in the mode-bar with the tab key. 
- Tab key navigation of the mode-bar links do not follow the order in which they are visually arranged on the page. 
- In devices with smaller width, outline around some links are hidden.

**Observation & steps taken:**
- Links having no href attribute were skipped. Added href attribute to the skipped links 
- Fixing this exposed that the order in which the tab key moves to the links is incorrect. 
- Made changes both in platform-client and platform-pattern library. Changed the arrangement of affected links in the html and styles. See the [associated  platform-pattern-library pull request](https://github.com/ushahidi/platform-pattern-library/pull/243).
- Outline hiding behind some of the mode-bar links in smaller devices not solved yet, thinking it's probably a bug in the browser for now.

**Testing checklist:**
- [x] Add assets folder from [platform-pattern-library containing the new changes](https://github.com/ushahidi/platform-pattern-library/pull/243/commits) made to the [platform-client which also contains some new changes](https://github.com/ushahidi/platform-client/pull/1564/commits) as instructed in documentation and run the project.
- [x] Login to your account.
- [x] Navigate to links in the mode-bar using the tab key
- [x] Verify that no link is skipped
- [x] Check that outline moves from one link to the next in the order that the links are visually arranged (also test backwards with shift + tab) 
- [x] Test after you log out from your account also. 

- [x] I certify that I ran my checklist

Fixes ushahidi/platform# .

Ping @ushahidi/platform

Ifycode commented:
[clabot:check]

Ifycode commented:
[clabot:check]

Ifycode commented:
Related change to styles can be found in this other pull request:
https://github.com/ushahidi/platform-pattern-library/pull/243

Angamanga commented:
@Ifycode Thank you! I will review shortly!

renujain31 commented on app/common/directives/mode-bar/mode-bar.html:
```diff
@@ -32,7 +32,7 @@
 
             <li>
-                <a class="more-menu-trigger" ng-class="{ active : moreActive }" ng-click="showMore($event)">
+                <a href="" class="more-menu-trigger" ng-class="{ active : moreActive }" ng-click="showMore($event)">
```
Hey @Ifycode I don't know if it is the right way to enable navigation. Can you see if we can do it  by setting the `tabindex` instead as it might be the suggested way to enable navigation.
You can read more about tabindex here: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

Maybe @Angamanga and @rowasc have more to add to this.

Angamanga replied:
@renujain31 Thank you noticing and reviewing 🎉

Ifycode replied:
Ok. Please see comment https://github.com/ushahidi/platform-client/pull/1564#issuecomment-713254498

Angamanga reviewed (changes requested):
Thank you for your contribution @Ifycode! This works fine together with the pattern-library-changes but could you change href="" to using tabindex instead as @renujain31 suggests?

Ifycode commented:
Ok @renujain31 @Angamanga Thanks for the review. I'm on it.

Ifycode commented:
Oh ok @renujain31 @Angamanga I'm using the tabIndex like you suggested on my local computer. I remember I have tried this before and that's why I resorted to using href="" instead. There is a problem with using tabIndex on the link tags <a> themselves (or even on the parent list li element). Usually after navigating the links with the tab key, the next thing for a keyboard input user is to use the enter key to access the pages they link to.

**Noticed the following problems with using tabIndex in this case:**
When using tabIndex="0", the affected links are focusable but the problem begins when you use the enter key to click on them - enter key doesn't respond. So the keyboard user is not able to access the pages they're meant to take them to, or open up a modal (in the case of those that don't lead to another page).

**Noticed the merit of using href in this case**
With the href, using the enter key makes the links both focusable and allows the user to access the pages that the links lead to or bring up the modal (in the case of modals).

You can try it on your end to see what I mean. I'm waiting to hear your response on this matter.

Ifycode commented:
Oh ok @renujain31 @Angamanga I'm using the tabIndex like you suggested on my local computer. I remember I have tried this before and that's why I resorted to using href="" instead. There is a problem with using tabIndex on the link tags <a> themselves (or even on the parent list li element). Usually after navigating the links with the tab key, the next thing for a keyboard input user is to use the enter key to access the pages they link to.

**Noticed the following problems with using tabIndex in this case:**
When using tabIndex="0", the affected links are focusable but the problem begins when you use the enter key to click on them - enter key doesn't respond. So the keyboard user is not able to access the pages they're meant to take them to, or open up a modal (in the case of those that don't lead to another page).

**Noticed the merit of using href in this case:**
With the href, the links are both focusable and using the enter key allows the user to access the pages that the links lead to or bring up the modal (in the case of modals).

You can try it on your end to see what I mean. I'm waiting to hear your response on this matter.

Angamanga commented:
@Ifycode  @renujain31 After discussing with @rowasc yesterday about how to handle this, we came to the conclusion that those last   <a>- tags (collections, sign in, sign-out, user-avatar, support and "more") are really to be considered as buttons and not anchor-tags. However, this requires changes in the pattern-library as well to make the styling look the same across the mode-bar. @Ifycode Do you think you could convert those a-tags to buttons and take a look at the pattern-library for styling as well? Sorry I did not realise this before you started!

renujain31 commented:
Ohkay @Angamanga, that makes sense. @Ifycode let me know if you face any difficulty I would be happy to help you.

Ifycode commented:
Ok @Angamanga i'll work on it. Thanks for offering to help out @renujain31.
If any questions or issues pop up, I'll sure let you know.

Ifycode commented:
Hi @Angamanga @renujain31, I have changed the links to buttons. Please help review.

Angamanga reviewed (approved):
This works fine now! Thank you for your contribution!

Angamanga commented:
@Ifycode This is approved but just saw there is a conflict in the mode-bar.html compared to develop , could you check and fix the conflicts? Thank you 🙏

Angamanga commented:
@Angamanga Wait for pl-release before QA!

Ifycode commented:
> @Ifycode This is approved but just saw there is a conflict in the mode-bar.html compared to develop , could you check and fix the conflicts? Thank you 🙏

Oh OK @Angamanga Will fix it.

Ifycode commented on app/common/directives/mode-bar/mode-bar.html:
```diff
@@ -72,6 +72,13 @@
+            <li ng-show="currentUser">
+                <a ng-click="viewAccountSettings()">
+                    <img class="avatar" ng-src="https://www.gravatar.com/avatar/{{ currentUser.gravatar || '00000000000000000000000000000000' }}?d=retro" alt="account_settings"/>
+                    <span class="label" translate="nav.account_settings">Account Settings</span>
```
Hi @Angamanga I need to clarify something, please help with this. 
- I got a bit confused while trying to resolve the conflict as you asked me to. Line 75-81 is newly added. Want to clarify if I should have totally removed it instead of adding it.

Ifycode replied:
- Another thing, I see that lines 71-81 and lines 92-97 are almost exactly the same. The difference between them is only **<span class="label" translate="nav.account_settings">Account Settings</span>** on line 79 and **<span class="label">{{currentUser.realname || currentUser.email}}</span>** on line 95 respectively.

Is line 75-81 meant to stay?

Angamanga replied:
Ok, yes there has been changes in the same code since you created this branch. So now we need to fix the code so it works with your changes. If you go to the browser now, you will see two icons with user-avatar when logged in. Line 76-81 ( https://github.com/ushahidi/platform-client/blob/1f0c99e191c6ec74c31f12a013a76b98ebf84633/app/common/directives/mode-bar/mode-bar.html#L76-L81) was made after you created the branch and line 92-97 (https://github.com/ushahidi/platform-client/blob/1f0c99e191c6ec74c31f12a013a76b98ebf84633/app/common/directives/mode-bar/mode-bar.html#L92-L97) was made by you. So what you need to do is adjust line 76-81 so it fits with your changes (changing the <a>-element to a <button> and after that remove line 92-97. You can do that in a new commit and push to this branch. Let me know if you need more guidance or want me to clarify something :)

Ifycode replied:
Oh. Ok. Doing this right away. Thanks for clarifying.

Ifycode replied:
> So what you need to do is adjust line 76-81 so it fits with your changes (changing the` <a>`-element to a `<button>`) and after that remove line 92-97. You can do that in a new commit and push to this branch. Let me know if you need more guidance or want me to clarify something :)

I have removed 92-97 and changed the a tag to button @Angamanga

Ifycode commented:
> So what you need to do is adjust line 76-81 so it fits with your changes (changing the` <a>`-element to a `<button>`) and after that remove line 92-97. You can do that in a new commit and push to this branch. Let me know if you need more guidance or want me to clarify something :)

I have removed 92-97 and changed the a tag to button @Angamanga

Angamanga reviewed (approved):
Thank you for your contribution 🎉

Angamanga commented:
@Obadha2 This is ready for QA. This pr has changes in the pattern-library too, I have created a PR for the new version here: https://github.com/ushahidi/platform-client/pull/1611. Merge that to develop before merging this one.

Angamanga commented:
@Obadha2 This pr needs this one to work: https://github.com/ushahidi/platform-client/pull/1611

Obadha2 commented:
QA'd, pases. 👍
